### DISCUSSION TITLE
Decorating proteins

### DISCUSSION THREAD

**Discussion by cesar-lopez-bau** [marrink-lab/polyply_1.0#363] [Ideas]
Hi guys;

Is there a way to attached modifications into a martinize2 generated topology. Let's say I want to attach CG glycans for which I can add the parameters into the polyply library. Then I would think of indicating the attachment in the protein and outputting a new structure + glycosylation while adding the new parameters into the original itp. I tried following the PEGtylated protein tutorial but it is somehow inefficient if I want to add more than 1 decoration and for many proteins. Any help will be appreciated. Cesar

**Comment by fgrunewald** [comment]
@cesar-lopez-bau you will probably not get around a little bit of scripting with python. The basic idea would be to create a .json file that describes your whole residue graph which is protein plus sugars. This graph needs to have each residue as a node and each node needs to have a `resid`, and `resname` attribute. In addition, the protein residues need to have a special attribute `"from_itp": "moleculetype"`, where moleculetype needs to be the same as in the itp file.

Let me give you an example. For the sake of brevity let's generate a peptide with polyply. Note that Step 1 in your case is then replaced by martinize2. 

===> Step 1
```
>PROTEIN sequence
AASAA
```

Now we generate a protein itp file using:

```
polyply gen_params -lib martini3 -seqf seq.fasta -o protein.itp -name protein
```
===> Step 2

Next we read in the fasta file and attache the PEG residues to the protein. The following code is an example on how to do this. Note that `MetaMolecule` is a `nx.Graph` object and has the same methods etc. 

```
import json
from pathlib import Path
import networkx as nx
from networkx.readwrite import json_graph
import vermouth
from vermouth.forcefield import ForceField
from polyply import MetaMolecule

ff = ForceField("test")
meta_mol = MetaMolecule.from_sequence_file(ff, Path("seq.fasta"), "protein")

# set node attribute from_itp
nx.set_node_attributes(meta_mol, "protein", "from_itp")

# add three residues of PEG or sugars
# !!!! You want to replace these three lines with some code that attaches the sugars !!!!
meta_mol.add_monomer(5, "PEO", [(2, 5)])
meta_mol.add_monomer(6, "PEO", [(5, 6)])
meta_mol.add_monomer(7, "PEO", [(6, 7)])

# networkx is abit restrictive for subclasses
# so we need to make a clean graph
seq_graph = nx.Graph()
seq_graph.add_nodes_from(meta_mol.nodes(data=True))
seq_graph.add_edges_from(meta_mol.edges)

# write out json file
g_json = json_graph.node_link_data(seq_graph)
with open("protein_plus_PEG.json", "w") as file_handle:
    json.dump(g_json, file_handle, indent=2)
```

==> Step 3: run the gen_params program

```
polyply gen_params -f PEO.martini3.ff protein.itp -seqf protein_plus_PEG.json -o combined.itp -name combined
```

Please note that the PEO.martini3.ff is modified for the example at hand:

```
[ moleculetype ]
; name nexcl.
PEO         1
;
[ atoms ]
1  SN3r    1   PEO   EC  1   0.000  45
;

[ link ]
resname "PEO"
[ atoms ]
EC { }
>EC { }
[ bonds ]
; back bone bonds
EC  >EC     1   0.36   7000 {"comment": "central-bond"}

[ link ]
resname "PEO"
[ dihedrals ]
EC >EC >>EC >>>EC  11 8 0.6570 -1.3278 -0.43661278 1.0808704 0.680055 {"ifndef": "TI"}

[ link ]
resname "PEO"
[ angles ]
EC  >EC >>EC    2   123.00  80 {"comment": "central-angle", "version":"1"}

[ link ]
resname "PEO|SER"
[ atoms ]
SC1 {"replace": {"atype": "TN3a"}}
>EC { }
[ bonds ]
SC1 >EC  2  0.34  1000
```

The coordinate generation works the same as in the PEGylated protein tutorial where you provide the coordinates of the CG protein to `polyply gen_coords` using the `-c` flag. 

If you have all the links the only thing to do is to build a custom script like the one above that specifies where to attach them to the protein. We are working with Paulo on a graphical user interface to do this automatically but it will take a few month until it is released. 

Let me know what you think about it and if you have any more questions!

Best,
Fabian
